File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    reverseList(head) {
        let reverseHead = null        
        let curr = head
        while(curr !== null){
            const prev = reverseHead
            reverseHead = new ListNode(curr.val, prev) 
            curr = curr.next
        }
        return reverseHead
    }
}

// A -> B -> C
// A <- B <- C

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # Time Complexity O(n): We traverse the list every single node
        # Space Complexity O(n): Every single node stored when no cycle 
        mem = {}
        curr = head 
        while curr:
            if curr in mem:
                return True
            mem[curr] = True
            curr = curr.next
        return False